### ISSUE TITLE
prisma generate --sql does not read env vars in AWS

### ISSUE THREAD

**Issue by filipjnc** [prisma/prisma#25512]
### Bug description

Today we introduced raw SQL queries in our codebase, requiring us to run `prisma generate --sql` instead of `prisma generate`. Locally it builds normally. In AWS, it errors out with `prisma generate --sql`:

```
#14 25.81 > ef-plugins-app@0.0.2 postinstall /app
--
4921 | #14 25.81 > prisma generate --sql
4922 | #14 25.81
4923 | #14 26.62 Prisma schema loaded from prisma/schema.prisma
4924 | #14 26.66 Error: Prisma schema validation - (get-config wasm)
4925 | #14 26.66 Error code: P1012
4926 | #14 26.66 error: Environment variable not found: DATABASE_URL.
4927 | #14 26.66   -->  prisma/schema.prisma:17
4928 | #14 26.66    \|
4929 | #14 26.66 16 \|   // https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#string
4930 | #14 26.66 17 \|   url      = env("DATABASE_URL")
4931 | #14 26.66    \|
4932 | #14 26.66
4933 | #14 26.66 Validation Error Count: 1
4934 | #14 26.66 [Context: getConfig]
```

When I omit `--sql`, it passes that build step but fails on the type checking later on, because we rely on the generated types.

### Prisma information

```prisma
generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native", "linux-musl-openssl-3.0.x"]
  previewFeatures = ["typedSql"]
}

datasource db {
  // provider = "postgresql"
  provider = "postgresql"
  // NOTE: When using mysql or sqlserver, uncomment the @db.Text annotations in model Account below
  // Further reading:
  // https://next-auth.js.org/adapters/prisma#create-the-prisma-schema
  // https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#string
  url      = env("DATABASE_URL")
}

generator zod {
  provider = "zod-prisma-types"
}

// Necessary for Next auth
model Account {
  id                String  @id @default(cuid())
  userId            String
  type              String
  provider          String
  providerAccountId String
  refresh_token     String? @db.Text
  access_token      String? @db.Text
  expires_at        Int?
  ext_expires_in    Int?
  token_type        String?
  scope             String?
  id_token          String? @db.Text
  session_state     String?
  user              User    @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([provider, providerAccountId])
}

model Session {
  id           String   @id @default(cuid())
  sessionToken String   @unique
  userId       String
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
  id                  String           @id @default(cuid())
  name                String?
  email               String           @unique
  emailVerified       DateTime?
  image               String?
  accepted            Boolean          @default(false)
  createdAt           DateTime         @default(now())
  updatedAt           DateTime         @updatedAt
  deletedAt           DateTime?
  roles               Role[]           @default([])
  entities            String[]         @default([])
  jobFamilies         String[]         @default([])
  jobClusters         String[]         @default([])
  jobFunctions        String[]         @default([])
  // Relationships
  accounts            Account[]
  sessions            Session[]
}
```

### Environment & setup

- OS: Linux
- Database: Postgres
- Node.js version: V20.16.0

### Prisma Version

```
5.21.1
```

**Comment by jkomyno** [comment]
Hi @filipjnc, where is the `DATABASE_URL` env variable defined? Is it in a local `.env` file, exported as a global variable, or something else? Thanks.

**Comment by mattkindy** [comment]
This seems like a breaking change that (as far as I can) aren't mentioned anywhere in the TypedSQL docs: namely, you have to be able to open a connection to the database when running `--sql`?

**Comment by giladv** [comment]
i want to revisit this issue. running generate --sql from a docker-compose process + env variable defined in the docker-compose file does not work. (aka Environment variable not found: DATABASE_URL.)
running the same thing without the --sql flag works just fine. something weird is happening here

**Comment by rstrand** [comment]
> i want to revisit this issue. running generate --sql from a docker-compose process + env variable defined in the docker-compose file does not work. (aka Environment variable not found: DATABASE_URL.) running the same thing without the --sql flag works just fine. something weird is happening here

@giladv its because the Typed SQL feature requires an active database connection
https://www.prisma.io/docs/orm/prisma-client/using-raw-sql/typedsql#active-database-connection-required

**Comment by kbzowski** [comment]
Actually active connection to the database is mentioned in the documentation, but probably poorly emphasized. 
Realistically I currently see two options for people using Docker:
- Have in DockerFile `ENV DATABASE_URL=` and provide access to the database
- Copy from host the content of  `node_modules/.prisma/client/sql` after generating the types (I haven't tried this).

Both strongly inconvenient. So what are the recommendations in building a Dockerfile with this option enabled? Because I must admit I like it a lot.

**Comment by dvlkv** [comment]
This is a significant problem to start using typedsql. We do not want our build infrastructure rely on database

**Comment by dvlkv** [comment]
> * Copy from host the content of  `node_modules/.prisma/client/sql` after generating the types (I haven't tried this).

This works as a workaround for me. Thanks!

**Comment by maciejopalinski** [comment]
One workaround is to output the generated Prisma Client code into some directory that is not ignored by your version control system, so CI/CD environment has access to types generated on local development environment.

But I hate this workaround, I can see at least 100 different ways how this can go wrong and out of sync.

**Comment by MorenoMdz** [comment]
Bumping this one, the tool is very interesting but the DB access is a blocker for us to use it in production.

**Comment by daw5** [comment]
Bump. This sucks!

**Comment by nikolas-beckel** [comment]
Bump.

**Comment by mysilkworld** [comment]
Bump.

**Comment by maciejopalinski** [comment]
I don't think it makes sense to bump this any further. The issue has been open for nearly a year with no indication that there's any intention to address it. In my experience, Prisma doesn't feel production-ready for anything beyond basic CRUD APIs. I attempted to use it in a production application, but it caused significant issues and ultimately forced me to migrate everything to a different ORM.

**Comment by buiducnhat** [comment]
Bump

**Comment by chriszrc** [comment]
I'm ok with the need for active database connection, but getting back to the OP issue, I have been unable to get the prisma generate --sql command to use env vars that are clearly defined on the system, but it refuses to use them. The only workaround I've found is to create a temporary .env file in the build directory, because that will work, then remove the .env file before the files get packaged for deployment-
